messaginginapp/src/sharedTest/java/io/customer/messaginginapp/GistSdkListenersTest.kt

```diff
@@ -0,0 +1,124 @@
+package io.customer.messaginginapp
+
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import io.customer.commontest.BaseTest
+import io.customer.messaginginapp.gist.data.model.Message
+import io.customer.messaginginapp.gist.presentation.GistListener
+import io.customer.messaginginapp.gist.presentation.GistSdk
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.TimeUnit
+import kotlin.concurrent.thread
+import org.amshove.kluent.internal.assertEquals
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+internal class GistSdkListenersTest : BaseTest() {
+    @Test
+    fun processAndRemoveListenersIndividually_givenConcurrentModification_expectSuccessfulCompletion() {
+        val listenersCount = 100
+        val emitEventsCount = listenersCount / 5
+        val listeners = ArrayList<GistListener>(listenersCount)
+        val threadsCompletionLatch = CountDownLatch(2)
+
+        // Add listeners
+        repeat(listenersCount) {
+            listeners.add(emptyGistListener())
+            GistSdk.addListener(listeners.last())
+        }
+
+        val removeListenersThread = thread(start = false) {
+            repeat(listenersCount) { index ->
+                GistSdk.removeListener(listeners[index])
+            }
+            threadsCompletionLatch.countDown()
+        }
+
+        val handleGistErrorThread = thread(start = false) {
+            repeat(emitEventsCount) {
+                GistSdk.handleGistError(Message())
+            }
+            threadsCompletionLatch.countDown()
+        }
+
+        handleGistErrorThread.start()
+        removeListenersThread.start()
+
+        // Wait for threads to complete without any exceptions within the timeout
+        // If there is any exception, the latch will not be decremented
+        threadsCompletionLatch.await(10, TimeUnit.SECONDS)
+
+        assertEquals(
+            expected = 0L,
+            actual = threadsCompletionLatch.count,
+            message = "Threads did not complete within the timeout"
+        )
+    }
+
+    @Test
+    fun processAndClearListenersAllAtOnce_givenConcurrentModification_expectSuccessfulCompletion() {
```

Comment:
Yes, because: 

- `removeListener` and `clearListeners` are two different functions in `GistSdk`
- They are independent and do not reuse each other
- They have different logics (one removes listener instantly while the latter loops over all listeners)

Reply:
Sorry, I didn't realize there were 2 different functions. Yes, having 2 test functions is good then.